### PR TITLE
Added an hybrid RayMarching/RayTracing mode for RTReflections and RTGI.

### PR DESCRIPTION
**Pull request by anisunity**
This PR Adds a new control mode for ray traced reflections and ray traced global illumination. This new mode combines both ray marching and ray tracing to solve visibility. This allows for both potiential peformance boost and access to details that cannot be stored in the RTAS (Decals or opaque particles). The option can be found on both volumes and are only compatible with the performance mode.

image

**Testing status**
Tested locally extensively with multiple scenes and assets.
Addeds 4 new tests to cover this new features.
QA needs to test because this changes quite a bit of code and is complex. Here are some things to test
- Verify that hybrid and ray tracing is the same (minus artefacts)
- Reflect sky should work the same way in both modes
- Test that supported ray tracing modes (in Asset) works as expected (Hybrid is supported only in both or perf mode)
- Test that hybrid mode works correctly on transparent that receives RTR
- Ensure that both parameters ray steps works as expected
- Test half resolution in hybrid
- Test Vertex Animation in hybrid

### CODE REVIEW

**Comment by 0lento** [comment]
When testing this on 2021.2.0a61011, I only see the hybrid mode doing ray marched reflections when using Custom quality mode. Low, Medium and High don't show ray marched reflections at all.

Repro:
- make new 3D project with 2021.2.0a6
- reference this branch packages to it
- run HDRP Wizard for HDRP + RT
- enable reflections from HDRP Asset
- add reflections to DXR volume
- add few primitives and vertex offsetting shader graph to visualize the change (that is not possible in RT reflections due to lack of vertex offset support)
- note how the only quality mode in reflection component that shows the dynamic reflection is Custom

**Comment by anisunity** [comment]
Fixed the issue reported by @0lento

**Review by JordanL8 — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Editor/RenderPipeline/Raytracing/SubSurfaceScatteringEditor.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ class SubSurfaceScatteringEditor : VolumeComponentEditor
 
         public override void OnEnable()
         {
-            var o = new PropertyFetcher<GlobalIllumination>(serializedObject);
+            var o = new PropertyFetcher<SubSurfaceScattering>(serializedObject);
```
[the remark this answers is not in the document]
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Editor/RenderPipeline/Raytracing/SubSurfaceScatteringEditor.cs:16]: Yes it is a fix, but let's do a diffrent PR for the backport

[the remark this answers is not in the document]
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Editor/RenderPipeline/Raytracing/SubSurfaceScatteringEditor.cs:16]: I have handled it, don't need to do separte PR (for all branch)

**Comment by sebastienlagarde** [comment]
PR waiting for QA review

**Comment by sebastienlagarde** [comment]
Note for myself: we have a problem with the force foward emissive features. As it will remove the Emissive contribution from the buffer it will not work with the remarching part of the algorithm.

**Review by remi-chapelain — CHANGES_REQUESTED** [review]
This is great and improves the range of what's possible with in HDRP

What has been tested : 
- Opaque particles 
- Hybrid reflection on transparent and with transparent objects
- Fallback if disabling ray tracing in the asset
- Fallback if changing ray tracing supported mode in the asset
- Decal reflection + Indirect lighting
- Reflection and GI with Vertex animation
- Half resolution + reflect sky

Here's a few things I noticed : 
1. For Transparent objects that receives reflection, hybrids needs to automatically fallback to "classic" ray tracing mode, to avoid this kind of things
https://user-images.githubusercontent.com/57442369/111299407-3040f080-8650-11eb-8f5f-b40cb8aa3eb7.mp4

2.  Currently, layer mask doesn't work with hybrid mode (with Indirect Lighting & Reflections)
https://user-images.githubusercontent.com/57442369/111299676-83b33e80-8650-11eb-9462-12eeb54f4d76.mp4

3. When switching between hybrid and ray tracing, the text where there's "Quality" and "Mode" slightly move and change size
https://user-images.githubusercontent.com/57442369/111299960-d987e680-8650-11eb-9cd1-39c047314bfc.mp4

4. Now that there's multiple mode and we use the word ray marching for classic SSR, we could change tooltip name for Max Ray Steps with "Ray Marching" instead of ray tracing
image

5. Tooltip for Ray Max Iterations RT is inexistent. And Since this is the same kind of parameters (same, but different) that the one in the previous point (#4), I suggest we keep the same kind of naming : "Max Hybrid Ray Steps", tooltip should be similar also
image

6. @JordanL8, since it's not especially trivial how hybrid mode work, I'd like to add some information about what's supported, what's not and why it's not with this mode to avoid users getting unexpected results. I got at least 3 things : 
- In the documentation, we kind of mention that the hybrid feature is only for opaque (opaque particle is used), because information are not entirely available for transparent since they are rendered in forward. In that sense, we should mention that it will work only for material rendered in deferred since it uses information in the GBuffer. 
- Also, It wouldn't hurt to emphasis a bit more that hybrid won't really help to get unsupported ray tracing feature (like decals.. etc) if the surface that receives reflection is transparent or if the reflected object is transparent because of forward.
- For Hybrid Raytraced GI, it can be interesting to note that we won't be able to get contribution from emissive decals because emission are computed later in the pipeline. For that same reason, as for now, we won't be able to get emissive contribution for materials with force forward emissive too.

**Comment by anisunity** [comment]
@remi-chapelain Fixed the following points 1 4 and 5

The point 2 is not a mistake, the layer mask still works for the ray traced bits.

The rest is doc.

@sebastienlagarde merged master

**Review by remi-chapelain — CHANGES_REQUESTED** [review]
1. Hybrid falls back to ray tracing when the receiver is transparent however, that fixes only half the issue and the issue introduced is a regression compared to master. In ray tracing perf mode (so in hybrid too), you lose the reflections for transparents objects that misses (fallback on sky) cf video

https://user-images.githubusercontent.com/57442369/112463599-bc46cc80-8d62-11eb-9131-87b17f11b252.mp4
 
2. Okay, so we can keep layer mask but I renamed the tooltip to say we use it for ray tracing (I noticed it was saying mask used for screen space reflections anyway.. so it was already wrong) ✔️
3.  Still having the issue but won't block for such minor/cosmetic issue ❌
4. All good ✔️ 
5. All good ✔️ 
6. Doc is up to date now ✔️

**Comment by Wenzy--** [comment]
In my testing. The instanced cylinder can reflect the diffuse color but not in the hybird mode. Is it normal ? 


Ray marching mode
[](https://ibb.co/ZJrQFVP)
Hybird mode
[](https://ibb.co/DtnQj5Y)

Also is it possible to improve the quality in Ray marching mode? The GI is very nice but the noise is a bit annoying. Especially in the shadow side

**Comment by landonth** [comment]
> 4\. Now that there's multiple mode and we use the word ray marching for classic SSR, we could change tooltip name for Max Ray Steps with "Ray Marching" instead of ray tracing

@remi-chapelain Can it also be considered that the ray model (Ray Marching, Ray Tracing, or some combination thereof) may not the most important distinction when setting expectations for the user here? 

1. My suggestion is that (currently) **the most important distinction** is whether the technique used is:
a. Able to access off-screen **World Space** scene information (here called "Ray Tracing")
-or-
b. Limited to accessing on-screen **Screen Space** scene information (here called "Ray Marching")
_- Because the user will immediately notice the visual difference in reflections, indirect lighting, etc taking  off-screen geometry and lights, etc into account (or not.)_

2. A close second in importance being the distinction between:
a. **Hardware accelerated** methods requiring newer GPU architecture such as RTX or RDNA2, providing significant performance gains over programmable GPU software methods alone.
-or-
b. **Software compute** methods, requiring only compute capable GPU architecture, (also perhaps when ray tracing procedurals in DXR / VKR as they can't take advantage of the triangle / ray intersection "RT cores" in dedicated GPU raytracing hardware and instead the work is handed off to conventional compute cores for rendering. @IonutNedelcuUnity mentioned something along these lines, anyway.)
_- Because the user will notice this when profiling performance_

3. The details of particular ray model(s) used coming in about third in importance to the user. I think it's still very helpful if the ray model used and other implementation details are mentioned somewhere in the UI, such as in tool-tips or in detail text, documentation, etc.
_- Because the user may never perceive the difference in ray model alone, as it is not mutually exclusive to either of the distinctions above, the ray model used in the implementation may also be subject to change without affecting qualitative visual results._

A few other related notes on naming conventions:
-  I don't think it applies here but just to make one exception to the above: even though it's considered a type of Ray Tracing, I believe Path Tracing is more deserving of a more prominent distinction as a unified ray / lighting model as it can more dramatically improve qualitative visual results, and greatly simplify workflows. I figure you all know this but just to make clear my position in context.
- Calling anything "Hybrid" in the context of render pipelines may create an ambiguity with the already vaguely named "Hybrid Renderer," (hybrid what? it should perhaps be in the name e.g. "Hybrid DOTS Renderer" Not sure how to best avoid or manage that here if at all. Just something else to keep in mind (I'm sure it is already for many.)

Thank you for your consideration and for all your work on these exciting new features ✨💜

**Comment by anisunity** [comment]
@sebastienlagarde @remi-chapelain Changes done

**Review by remi-chapelain — COMMENTED** [review]
Committed a few changes : 
- "Hybrid" becomes now "Mixed" to avoid confusion as @landonth pointed
- Tooltips have been improved with information about the tracing mode (hardware, screen/world space)
- Renamed a few variables for consistence between override

Also on the QA side : 
- Checked the last fix
- Verified that everything was properly noted on the docs

There's one last things that I want to make sure before approving : 

1. Why do we have the max ray steps and max mixed ray steps in a open field in the SSR override and we have them in a slider (with a max) in the SSGI override ? Is there a reason, if not, could we make that consistent ? 

image

**Review by remi-chapelain — APPROVED** [review]
What has been fixed in the last commits (pair debugging with @anisunity)
- Changed the user facing displayed value "Max Mixed Ray Steps" to "Max Ray Steps" to avoid having to confuse further users. The variable has the same naming for "Mixed Tracing" and "Ray Marching", but it's not the same variable behind; The tooltips explain that. 
- Fixed a few issues regarding Quality settings depending on which is supported is in the assets and the tracing mode selected
- Unclamped the max ray steps parameters in the GI override to match what is done in the screen space reflection override. 
- Changed max ray steps parameters for mixed tracing to match medium default quality settings

Ready for merge ✔️

**Review by remi-chapelain — APPROVED** [review]

**Pull request merged**